### PR TITLE
[DP-402] spark - Riak Connection Pool - Introduce Inactivity timeout based eviction

### PR DESCRIPTION
**Pull request by zkhadikova**
WIP Do not merge

### CODE REVIEW

**Comment by srgg** [comment]
inactivityTimeout should be described in readme.md https://github.com/basho/spark-riak-connector#configuration

File: connector/src/main/scala/com/basho/riak/spark/rdd/connector/SessionCache.scala [hunk 1]
```diff
@@ -113,6 +118,56 @@ class SessionCache(val cacheSize: Int) extends Logging {
...
+
+  private val checkForExpiredSessionsTask = new Runnable() {
+    override def run(): Unit = {
+      checkForExpiredSessions
+    }
+  }
+
+  private val scheduledExecutorService = Executors.newSingleThreadScheduledExecutor(new ThreadFactory() {
+    override def newThread(r: Runnable) = {
+      val thread = Executors.defaultThreadFactory.newThread(r)
```
Comment (srgg) [inline connector/src/main/scala/com/basho/riak/spark/rdd/connector/SessionCache.scala]: Please set corresponding name for this thread.

File: connector/src/main/scala/com/basho/riak/spark/rdd/connector/SessionCache.scala [hunk 2]
```diff
@@ -113,6 +118,56 @@ class SessionCache(val cacheSize: Int) extends Logging {
...
+        ReleaseTask(session).run
+  }
+
+  private val checkForExpiredSessionsTask = new Runnable() {
+    override def run(): Unit = {
+      checkForExpiredSessions
+    }
+  }
+
+  private val scheduledExecutorService = Executors.newSingleThreadScheduledExecutor(new ThreadFactory() {
```
Comment (srgg) [inline connector/src/main/scala/com/basho/riak/spark/rdd/connector/SessionCache.scala]: BTW why don't us get rid of this executor and use Runnable instead

File: connector/src/main/scala/com/basho/riak/spark/rdd/connector/SessionCache.scala [hunk 3]
```diff
@@ -198,7 +255,11 @@ class CachedSession(conf: RiakConnectorConf, riakClient: RiakClient) extends Ria
       logDebug(s"Riak CachedSession was destroyed/shutdown: $conf")
       closed = true
     }
+    if (1 == refCount.get) {
```
Comment (srgg) [inline connector/src/main/scala/com/basho/riak/spark/rdd/connector/SessionCache.scala]: better to rewrite as:

```
    refCount.decrementAndGet() match {
      case 0 =>
        unwrap().shutdown()
        logDebug(s"Riak CachedSession was destroyed/shutdown: $conf")
        closed = true

      case 1 =>
        afterClose(this)

      case _ => // Nothing to do
    }
```

File: connector/src/main/scala/com/basho/riak/spark/rdd/connector/SessionCache.scala (hunk 2 above)
Comment (zkhadikova) [inline connector/src/main/scala/com/basho/riak/spark/rdd/connector/SessionCache.scala]: I just prefer ScheduledExecutor since it's self explanatory

Comment (srgg) [inline connector/src/main/scala/com/basho/riak/spark/rdd/connector/SessionCache.scala]: Sorry but from my PoV it isn't.. You are using it for running 1 recurring task in the one thread..

Therefore for me personally, in this particular ca, having the overall processing cycle in the one place is more clearly:
```
() -> {
// Setup thread name
// Do something 
sleep();
}
```

File: connector/src/main/scala/com/basho/riak/spark/rdd/connector/SessionCache.scala [hunk 4]
```diff
@@ -113,6 +119,62 @@ class SessionCache(val cacheSize: Int) extends Logging {
...
+              case e: Exception => logError(s"Unexpected exception while releasing expired session: ${session.conf}", e)
+            }
+          }
+        }
+        try {
+          Thread.sleep(checkRate);
+        } catch {
+          case e: InterruptedException => {
+            logWarning(s"Session Cleaner was interrunpted", e)
+            interrupt()
```
Comment (srgg) [inline connector/src/main/scala/com/basho/riak/spark/rdd/connector/SessionCache.scala]: Could you please change it to using 'break'?

**Pull request merged**